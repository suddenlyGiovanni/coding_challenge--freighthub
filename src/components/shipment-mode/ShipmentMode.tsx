import * as React from 'react'

import FormControl from '@material-ui/core/FormControl'
import FormGroup from '@material-ui/core/FormGroup'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import Checkbox from '@material-ui/core/Checkbox'

import { TransportMode } from 'components'

interface State {
  sea: boolean
  air: boolean
  rail: boolean
}

export interface Props {
  initialValues?: State
  onChange: (values: State) => void
}
export const ShipmentMode: React.FC<Props> = ({
  initialValues = { sea: true, air: true, rail: true },
  onChange,
}) => {
  const [state, setState] = React.useState<State>(initialValues)

  const handleChange = (name: keyof State) => (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const stateSnapshot = { ...state, [name]: event.target.checked }
    setState(stateSnapshot)
    onChange(stateSnapshot)
  }

  const { sea, air, rail } = state

  return (
    <FormControl component="fieldset">
      <FormGroup style={{ flexDirection: 'row' }}>
        <FormControlLabel
          control={
            <Checkbox
              checked={sea}
              onChange={handleChange('sea')}
              value="sea"
              icon={<TransportMode checked={false} mode={'sea'} />}
              checkedIcon={<TransportMode checked={true} mode={'sea'} />}
            />
          }
          label="Sea"
        />
        <FormControlLabel
          control={
            <Checkbox
              checked={air}
              onChange={handleChange('air')}
              value="air"
              icon={<TransportMode checked={false} mode={'air'} />}
              checkedIcon={<TransportMode checked={true} mode={'air'} />}
            />
          }
          label="Air"
        />
        <FormControlLabel
          control={
            <Checkbox
              checked={rail}
              onChange={handleChange('rail')}
              value="rail"
              icon={<TransportMode checked={false} mode={'rail'} />}
              checkedIcon={<TransportMode checked={true} mode={'rail'} />}
            />
          }
          label="Rail"
        />
      </FormGroup>
    </FormControl>
  )
}
