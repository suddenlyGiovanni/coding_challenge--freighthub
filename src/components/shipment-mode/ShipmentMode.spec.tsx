import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import { ShipmentMode, Props } from './ShipmentMode'

describe('ShipmentMode', () => {
  const props: Props = {
    initialValues: {
      sea: true,
      air: true,
      rail: true,
    },
    onChange: values => {},
  }

  test('snapshot', () => {
    // Arrange
    const { container } = render(<ShipmentMode {...props} />)
    // Assert
    expect(container).toMatchSnapshot()
  })

  it('calls `onChange` callback when the user clicks on the icons', () => {
    // Arrange
    const handleOnChange = jest.fn()
    const { getByLabelText } = render(
      <ShipmentMode {...props} onChange={handleOnChange} />
    )
    const seaCheckboxElement = getByLabelText('Sea')
    const railCheckboxElement = getByLabelText('Rail')
    const airCheckboxElement = getByLabelText('Air')

    // // Act
    fireEvent.click(seaCheckboxElement)
    fireEvent.click(railCheckboxElement)
    fireEvent.click(airCheckboxElement)

    // // Assert
    expect(handleOnChange).toHaveBeenCalledTimes(3)
    expect(handleOnChange).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({ sea: false })
    )
    expect(handleOnChange).toHaveBeenNthCalledWith(
      2,
      expect.objectContaining({ rail: false })
    )
    expect(handleOnChange).toHaveBeenNthCalledWith(
      3,
      expect.objectContaining({ air: false })
    )
  })
})
