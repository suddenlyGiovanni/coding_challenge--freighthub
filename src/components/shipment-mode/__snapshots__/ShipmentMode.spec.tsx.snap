// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShipmentMode snapshot 1`] = `
<div>
  <fieldset
    class="MuiFormControl-root"
  >
    <div
      class="MuiFormGroup-root"
      style="flex-direction: row;"
    >
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-19 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-20 Mui-checked MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
              data-testid="ship-icon"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M20 21c-1.39 0-2.78-.47-4-1.32-2.44 1.71-5.56 1.71-8 0C6.78 20.53 5.39 21 4 21H2v2h2c1.38 0 2.74-.35 4-.99 2.52 1.29 5.48 1.29 8 0 1.26.65 2.62.99 4 .99h2v-2h-2zM3.95 19H4c1.6 0 3.02-.88 4-2 .98 1.12 2.4 2 4 2s3.02-.88 4-2c.98 1.12 2.4 2 4 2h.05l1.89-6.68c.08-.26.06-.54-.06-.78s-.34-.42-.6-.5L20 10.62V6c0-1.1-.9-2-2-2h-3V1H9v3H6c-1.1 0-2 .9-2 2v4.62l-1.29.42c-.26.08-.48.26-.6.5s-.15.52-.06.78L3.95 19zM6 6h12v3.97L12 8 6 9.97V6z"
              />
            </svg>
            <input
              checked=""
              class="PrivateSwitchBase-input-22"
              data-indeterminate="false"
              type="checkbox"
              value="sea"
            />
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Sea
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-19 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-20 Mui-checked MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
              data-testid="plane-icon"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M21 16v-2l-8-5V3.5c0-.83-.67-1.5-1.5-1.5S10 2.67 10 3.5V9l-8 5v2l8-2.5V19l-2 1.5V22l3.5-1 3.5 1v-1.5L13 19v-5.5l8 2.5z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
            <input
              checked=""
              class="PrivateSwitchBase-input-22"
              data-indeterminate="false"
              type="checkbox"
              value="air"
            />
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Air
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-19 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-20 Mui-checked MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
              data-testid="train-icon"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2c-4 0-8 .5-8 4v9.5C4 17.43 5.57 19 7.5 19L6 20.5v.5h2.23l2-2H14l2 2h2v-.5L16.5 19c1.93 0 3.5-1.57 3.5-3.5V6c0-3.5-3.58-4-8-4zM7.5 17c-.83 0-1.5-.67-1.5-1.5S6.67 14 7.5 14s1.5.67 1.5 1.5S8.33 17 7.5 17zm3.5-7H6V6h5v4zm2 0V6h5v4h-5zm3.5 7c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
              />
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
            </svg>
            <input
              checked=""
              class="PrivateSwitchBase-input-22"
              data-indeterminate="false"
              type="checkbox"
              value="rail"
            />
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Rail
        </span>
      </label>
    </div>
  </fieldset>
</div>
`;
